### PR TITLE
feat: Add reports API endpoints [TA-2544]

### PR DESCRIPTION
**Pull request by szeying-lee**
# Description

This PR adds documentation for the reports API endpoints. Note that this PR does not include any of the fields used for recurring reports as that will be added in a later ticket.

Variants of the word `execution` is also added to `.alexrc.json` because of this warning:
```
Be careful with `executions`, it’s profane in some cases
```

Fixes # (issue)
[TA-2544](https://gr4vy.atlassian.net/browse/TA-2544)

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own changes
- [ ] I have run `yarn test` to make sure my changes pass all linters
- [ ] I have pulled the latest changes from the upstream `main` branch

## Contribution guidelines

For contribution guidelines, styleguide, and other helpful information please
see the `CONTRIBUTING.md` file in the root of this project.

### CODE REVIEW

File: reference/openapi.yaml [hunk 1]
```diff
@@ -157,6 +157,21 @@ tags:
...
+      the API and executed on a one-off or recurring basis.
+
+      The reports API can be used to:
+
+      * Create and configure new reports.
+      * List all reports.
+      * View the configuration details of a report.
+      * List a report's executions.
+      * Reconfigure an existing report.
+      * Generate a temporary URL to download a report's execution.
```
Comment (victormferrara) [inline reference/openapi.yaml:173]:
```suggestion
      * Generate a temporary URL to download a report result in csv format.
```
Maybe?

File: reference/paths/reports__list-report-executions.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+---
+operationId: list-reports-executions
```
Comment (victormferrara) [inline reference/paths/reports__list-report-executions.yaml:2]:
```suggestion
operationId: list-report-executions
```

File: reference/paths/reports__add-report.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      application/json:
+        schema:
+          $ref: "../openapi.yaml#/components/schemas/Error401Unauthorized"
+
+  "409":
+    description: Returns an error if duplicate resource has been found.
+    content:
+      application/json:
+        schema:
+          $ref: "../openapi.yaml#/components/schemas/Error409DuplicateRecord"
```
Comment (victormferrara) [inline reference/paths/reports__add-report.yaml:49]: Can this really happen? I can't find any related code, although might not be explicit.

File: reference/resources/report_executions.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+---
+title: Report executions
+type: object
+description: A list of executions for a report.
+x-tags:
+  - Reports
+
+properties:
```
Comment (victormferrara) [inline reference/resources/report_executions.yaml:8]: Isn't there also information about the limit and cursors? Same for `reports.yaml`.

File: reference/resources/report_spec.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+---
+title: Report Specification
+type: object
+description: The specification of a report.
+x-tags:
+  - Reports
+
```
Comment (victormferrara) [inline reference/resources/report_spec.yaml:7]: Should we add required fields here? I see that normally they're not specified for resources returned by the server, but this one is also received by the server when creating reports.

File: reference/resources/report_spec.yaml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+      - transactions
+
+  params:
+    type: object
+    description: |-
+      Parameters used to configure the report.
+    additionalProperties:
+      oneOf:
+        - type: array
+        - type: object
```
Comment (victormferrara) [inline reference/resources/report_spec.yaml:25]: Should we try to define the concrete expectations for a transactions report? It's going to be challenging though.

File: reference/resources/report_spec.yaml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+description: The specification of a report.
+x-tags:
+  - Reports
+
+properties:
+  model:
+    type: string
+    description: |-
+      The model (resource) that the data used for the report is retrieved
+      from.
```
Comment (federicojasson) [inline reference/resources/report_spec.yaml:13]: I think we may eventually have models that don't relate to only one resource, perhaps we can rephrase this to indicate that. A model as a "dataset".

File: reference/resources/report_spec.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    example: transactions
+    enum:
+      - transactions
+
+  params:
+    type: object
+    description: |-
+      Parameters used to configure the report.
```
Comment (federicojasson) [inline reference/resources/report_spec.yaml:21]: Should we clarify that the acceptable values here depend on the `model`?

File: reference/resources/report_execution_url.yaml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    format: uri
+    description: |-
+      The URL to download the report execution.
+    example: https://example.com/download
+
+  expires_at:
+    type: string
+    description: |-
+      The date and time that the signed URL to download the
+      report execution will expire.
```
Comment (federicojasson) [inline reference/resources/report_execution_url.yaml:20]: We can probably simplify this description as "The date and time when the URL expires". WDYT?

File: reference/resources/report.yaml [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  spec:
+    allOf:
+      - $ref: "../openapi.yaml#/components/schemas/ReportSpec"
+      - description: |-
+          The specifications of this report.
+
+  latest_execution:
+    allOf:
+      - $ref: "../openapi.yaml#/components/schemas/ReportExecution"
+      - description: |-
```
Comment (federicojasson) [inline reference/resources/report.yaml:60]: Is this description shown anywhere? I cannot find it :thinking:

**Review by federicojasson — APPROVED** [review]

File: reference/resources/report_spec.yaml (hunk 6 above)
Comment (szeying-lee) [reply reference/resources/report_spec.yaml:26]: I've left it out of this PR as I think this will be covered by [TA-2712](https://gr4vy.atlassian.net/browse/TA-2712), so for now it's just a very simple example.

File: reference/resources/report_executions.yaml (hunk 4 above)
Comment (szeying-lee) [reply reference/resources/report_executions.yaml:8]: Good spot, fixed in https://github.com/gr4vy/gr4vy-openapi/pull/218/commits/44aed44bd30506ba5bc4dfb505a1b80ffe188b5d

File: reference/paths/reports__add-report.yaml (hunk 3 above)
Comment (szeying-lee) [reply reference/paths/reports__add-report.yaml]: You're right, I've removed it, it was a bad copy and paste from another resource 😅 https://github.com/gr4vy/gr4vy-openapi/pull/218/commits/d50957dc2ca2646b12924d35f4f8837da40647b3

File: reference/openapi.yaml [hunk 11]
```diff
@@ -157,6 +157,21 @@ tags:
       * Provide Gr4vy with the payment credentials for a payment provider.
       * Set a display name for a payment provider.
 
+  - name: Reports
+    description: |-
+      In Gr4vy, a report contains a set of information about a specific type
+      of Gr4vy resource such as transactions. Reports may be configured via
+      the API and executed on a one-off or recurring basis.
```
[the remark this answers is not in the document]
Comment (szeying-lee) [reply reference/openapi.yaml]: I reworded quite a bit of the descriptions here: https://github.com/gr4vy/gr4vy-openapi/pull/218/commits/353357469c7f79b9509f2e68c77b3cbf0e2bc763

File: reference/openapi.yaml (hunk 1 above)
Comment (szeying-lee) [reply reference/openapi.yaml]: Makes more sense, I've reworded it further to fit with the rest of the descriptions: https://github.com/gr4vy/gr4vy-openapi/pull/218/commits/353357469c7f79b9509f2e68c77b3cbf0e2bc763

File: reference/resources/report_spec.yaml (hunk 5 above)
Comment (szeying-lee) [reply reference/resources/report_spec.yaml:7]: Makes sense - done in: https://github.com/gr4vy/gr4vy-openapi/pull/218/commits/12fb73afc33bb4508cf86ec3eb73413a37174e11

**Review by victormferrara — APPROVED** [review]
